### PR TITLE
DATACMNS-650 - Provide a CloseableIterator abstraction as a foundation for streaming of results.

### PR DESCRIPTION
**Pull request by thomasdarimont**
Introduced CloseableIterator, AutoCloseableIterator and ForwardingAutoCloseableIterator to bridge between the two.

### CODE REVIEW

**Comment by jhoeller** [comment]
Guys, java.io.Closeable extends java.lang.AutoCloseable anyway... So I'm not sure what AutoCloseableIterator really adds here, and why the forwarding is needed?

We're implementing Java 6's Closeable to implicitly get Java 7 AutoCloseable behavior in a few places in the core framework as well. Seems to work fine for those purposes.

**Comment by thomasdarimont** [comment]
`AutoCloseable` was just introduced with Java 7 so it isn't available if SD Commons is used with Java 6.

**Comment by jhoeller** [comment]
Sure, but my point is: Any of our classes implementing Closeable implicitly implements AutoCloseable on Java 7, since the Closeable interface itself happens to extend AutoCloseable on Java 7+. This is perfectly Java 6 compatible and also enables AutoCloseable when running on Java 7, simply through implementing Closeable. Any extra business with AutoCloseable decoration is effectively unnecessary.

**Comment by thomasdarimont** [comment]
A gotcha :) You are right.

File: src/main/java/org/springframework/data/util/ReflectionUtils.java [hunk 1]
```diff
@@ -30,10 +31,26 @@
...
  */
 public abstract class ReflectionUtils {
 
+	private static final Class<?> JAVA8_STREAM_TYPE;
+
+	static {
+		String streamClassName = "java.util.stream.Stream";
+
+		Class<?> cls = null;
+		if (ClassUtils.isPresent(streamClassName, ReflectionUtils.class.getClassLoader())) {
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/util/ReflectionUtils.java]: Can't we just use `ClassUtils.forName(…)` here?

**Comment by olivergierke** [comment]
As outline in the comments on the [original pull request](https://github.com/spring-projects/spring-data-mongodb/pull/274#issuecomment-75436060) I think we should focus on a `CloseableIterable` as return type. Using an `Iterator` as return type seems rather uncommon.

**Comment by thomasdarimont** [comment]
After having played a bit with the API I still think that a `CloseableIterable<T>`provides a richer API experience.

Consider the following variants:
```java
//Variant 1 - CloseableIterator<Person>
@Query("{firstname:{$in:?0}}")
CloseableIterator<Person> findByCustomQueryByFirstnames(List<String> firstnames);

//V1: Use case 1 - loop over all results.
try (CloseableIterator<Person> result = repository.findByCustomQueryByFirstnames(Arrays.asList("Dave"))){
	for (Person person : result) {
		readPersons.add(person);
	}
}

//V1: Use case 2 - check for results and process later
try (CloseableIterator<Person> result = repository.findByCustomQueryByFirstnames(Arrays.asList("Dave"))){
	if(result.hasNext()){
		//doSomething
	}else{
		//doSomething else
	}

	for (Person person : result) {
		readPersons.add(person);
	}	
}


//Variant 2 - CloseableIterable<Person>
@Query("{firstname:{$in:?0}}")
CloseableIterable<Person> findByCustomQueryByFirstnames(List<String> firstnames);

//V2: Use case 1 - loop over all results.
try (CloseableIterable<Person> result = repository.findByCustomQueryByFirstnames(Arrays.asList("Dave"))){
	for (Person person : result) {
		readPersons.add(person);
	}
}

//V2: Use case 2 - check for results and process later
try (CloseableIterable<Person> result = repository.findByCustomQueryByFirstnames(Arrays.asList("Dave"))){

    Iterator iter = result.iterator();

	if(iter.hasNext()){
		//doSomething
	} else{
		//doSomething else
	}

	for (Person person : result) {
		readPersons.add(person);
	}	

	//what to close now? iterator or result?
}
````
In Variant1 it is clear what the "resource" and a user can conveniently iterate over the results.
Variant2 however doesn't make this clear per se - IMHO.

**Pull request closed without merging**